### PR TITLE
Fixed a bug which makes Souper prefers system z3 on mac

### PR DESCRIPTION
**Pull request by zhengyangl**
According to https://cmake.org/cmake/help/v3.0/command/find_library.html , the search path may include directories other than specified ones if NO_DEFAULT_PATH is not set.

### CODE REVIEW

**Comment by manasij7479** [comment]
Does
```DYLD_PRINT_LIBRARIES=YES ./alive ../tests/unit/cttz.opt```
show it linked to the correct library?

**Comment by regehr** [comment]
I was using that test when trying to fix alive, but since this patch is hacking on souper, the test should be to do same for souper pass, souper-check, etc.

**Comment by regehr** [comment]
can you build this PR on your mac and check, @manasij7479 ?

**Comment by zhengyangl** [comment]
Alive2 should be patched separately, 

```
dyld: loaded: /Users/liuz/work/souper/third_party/alive2/build/./alive
dyld: loaded: /usr/local/lib/libz3.dylib
dyld: loaded: /usr/lib/libc++.1.dylib
dyld: loaded: /usr/lib/libSystem.B.dylib
dyld: loaded: /usr/lib/libc++abi.dylib
dyld: loaded: /usr/lib/system/libcache.dylib
dyld: loaded: /usr/lib/system/libcommonCrypto.dylib
dyld: loaded: /usr/lib/system/libcompiler_rt.dylib
dyld: loaded: /usr/lib/system/libcopyfile.dylib
dyld: loaded: /usr/lib/system/libcorecrypto.dylib
dyld: loaded: /usr/lib/system/libdispatch.dylib
dyld: loaded: /usr/lib/system/libdyld.dylib
dyld: loaded: /usr/lib/system/libkeymgr.dylib
dyld: loaded: /usr/lib/system/liblaunch.dylib
dyld: loaded: /usr/lib/system/libmacho.dylib
dyld: loaded: /usr/lib/system/libquarantine.dylib
dyld: loaded: /usr/lib/system/libremovefile.dylib
dyld: loaded: /usr/lib/system/libsystem_asl.dylib
dyld: loaded: /usr/lib/system/libsystem_blocks.dylib
dyld: loaded: /usr/lib/system/libsystem_c.dylib
dyld: loaded: /usr/lib/system/libsystem_configuration.dylib
dyld: loaded: /usr/lib/system/libsystem_coreservices.dylib
dyld: loaded: /usr/lib/system/libsystem_darwin.dylib
dyld: loaded: /usr/lib/system/libsystem_dnssd.dylib
dyld: loaded: /usr/lib/system/libsystem_info.dylib
dyld: loaded: /usr/lib/system/libsystem_m.dylib
dyld: loaded: /usr/lib/system/libsystem_malloc.dylib
dyld: loaded: /usr/lib/system/libsystem_networkextension.dylib
dyld: loaded: /usr/lib/system/libsystem_notify.dylib
dyld: loaded: /usr/lib/system/libsystem_sandbox.dylib
dyld: loaded: /usr/lib/system/libsystem_secinit.dylib
dyld: loaded: /usr/lib/system/libsystem_kernel.dylib
dyld: loaded: /usr/lib/system/libsystem_platform.dylib
dyld: loaded: /usr/lib/system/libsystem_pthread.dylib
dyld: loaded: /usr/lib/system/libsystem_symptoms.dylib
dyld: loaded: /usr/lib/system/libsystem_trace.dylib
dyld: loaded: /usr/lib/system/libunwind.dylib
dyld: loaded: /usr/lib/system/libxpc.dylib
dyld: loaded: /usr/lib/libobjc.A.dylib
Processing ../tests/unit/cttz.opt..
Couldn't read the file
```

**Comment by zhengyangl** [comment]
Wait the Z3_LIBRARY string becomes empty

**Comment by regehr** [comment]
I built this branch and souper-check fails to grab the internal z3

```Johns-MacBook-Pro:build johnregehr$ otool -L souper-check 
souper-check:
	@rpath/libalive2.dylib (compatibility version 0.0.0, current version 0.0.0)
	libz3.dylib (compatibility version 0.0.0, current version 0.0.0)
	/usr/lib/libz.1.dylib (compatibility version 1.0.0, current version 1.2.11)
	/usr/lib/libncurses.5.4.dylib (compatibility version 5.4.0, current version 5.4.0)
	/usr/lib/libSystem.B.dylib (compatibility version 1.0.0, current version 1252.200.5)
	/usr/lib/libxml2.2.dylib (compatibility version 10.0.0, current version 10.9.0)
	@rpath/libc++.1.dylib (compatibility version 1.0.0, current version 1.0.0)
Johns-MacBook-Pro:build johnregehr$ 
```

**Comment by regehr** [comment]
can you share the patch?

**Comment by zhengyangl** [comment]
Check https://github.com/zhengyangl/alive2/commits/fix

**Comment by regehr** [comment]
thanks! but we can't hard-code this path into alive, it has to come in through alive cmake

**Comment by zhengyangl** [comment]
No luck, alive2 is still trying to link agains system z3 after the patch.

File: CMakeLists.txt [hunk 1]
```diff
@@ -123,28 +124,31 @@ find_library(HIREDIS_LIBRARY
   PATHS
   third_party/hiredis/install/lib)
 
-find_library(ALIVE_LIBRARY alive2 PATHS "${ALIVE_DIR}/build")
+find_library(ALIVE_LIBRARY alive2 PATHS "${ALIVE_DIR}/build" NO_DEFAULT_PATHS)
```
Comment (regehr) [inline CMakeLists.txt]: I can see cmake documentation for NO_DEFAULT_PATH but not NO_DEFAULT_PATHS, can you point me to some documentation about this?

Comment (zhengyangl) [reply CMakeLists.txt]: Sorry that was typos.

Comment (regehr) [reply CMakeLists.txt]: do you think those are needed? do they help?

**Comment by zhengyangl** [comment]
Got this error when linking Alive2 statically against z3 on linux
```
FAILED: alive
: && /usr/bin/c++  -Wall -march=native -fPIC  -O3 -DNDEBUG -O3 -DNDEBUG   CMakeFiles/alive.dir/tools/alive.cpp.o CMakeFiles/alive.dir/tools/alive_lexer.cpp.o CMakeFiles/alive.dir/tools/alive_parser.cpp.o  -o alive  libir.a libsmt.a libtools.a libutil.a /home/liuz/work/souper/third_party/z3-install/lib/libz3.a && :
/home/liuz/work/souper/third_party/z3-install/lib/libz3.a(parallel_tactic.o): In function `parallel_tactic::operator()(ref<goal> const&, sref_buffer<goal, 16u>&)':
parallel_tactic.cpp:(.text._ZN15parallel_tacticclERK3refI4goalER11sref_bufferIS1_Lj16EE[_ZN15parallel_tacticclERK3refI4goalER11sref_bufferIS1_Lj16EE]+0x442): undefined reference to `pthread_create'
/home/liuz/work/souper/third_party/z3-install/lib/libz3.a(scoped_timer.o): In function `scoped_timer::~scoped_timer()':
scoped_timer.cpp:(.text+0xa1): undefined reference to `pthread_join'
/home/liuz/work/souper/third_party/z3-install/lib/libz3.a(scoped_timer.o): In function `scoped_timer::imp::imp(unsigned int, event_handler*)':
scoped_timer.cpp:(.text._ZN12scoped_timer3impC2EjP13event_handler[_ZN12scoped_timer3impC5EjP13event_handler]+0x45): undefined reference to `pthread_create'
collect2: error: ld returned 1 exit status
```

**Comment by regehr** [comment]
the -pthread option fixes this

**Comment by regehr** [comment]
glad you're testing on linux, I'll wait to merge until you've tested everything on both platforms

File: CMakeLists.txt (hunk 1 above)
Comment (zhengyangl) [reply CMakeLists.txt]: This option forces the find path, this may prevent potential library lookup mistake in the future.

**Comment by zhengyangl** [comment]
commit id of alive2 needed to be update, after the last PR to alive2 get merged.

**Comment by zhengyangl** [comment]
Ready for review, tested good on linux and mac. On linux I tested with gcc for both debug and release build. On mac I tested with clang for both debug and release build. There are several assertion failure triggered on debug builds, which are irrelevant to the build infrastructure. I'll send another PR for the assertion fix.

File: CMakeLists.txt [hunk 2]
```diff
@@ -123,28 +124,31 @@ find_library(HIREDIS_LIBRARY
...
+  third_party/z3-install/lib
+  NO_DEFAULT_PATHS)
 
 find_program(Z3
   NAMES
   z3
   PATHS
-  third_party/z3-install/bin)
+  third_party/z3-install/bin
+  NO_DEFAULT_PATHS)
```
Comment (regehr) [inline CMakeLists.txt]: I looked in the CMake source code and there's no support for something called NO_DEFAULT_PATHS

Comment (regehr) [reply CMakeLists.txt]: I think it should be NO_DEFAULT_PATH like you had before

**Comment by regehr** [comment]
LGTM, thanks!

**Pull request merged**